src/compiler/declarationEmitter.ts

```diff
@@ -133,6 +176,419 @@ namespace ts {
             referencePathsOutput,
         };
 
+        function emitReferences(sourceFile: SourceFile, emittedReferencedFiles: FileReference[]) {
+            // Check what references need to be added
+            if (!compilerOptions.noResolve) {
+                forEach(sourceFile.referencedFiles, fileReference => {
+                    let referencedFile = tryResolveScriptReference(host, sourceFile, fileReference);
+
+                    // If the reference file is a declaration file or an external module, emit that reference
+                    if (referencedFile && (isExternalModuleOrDeclarationFile(referencedFile) &&

```

Comment:
We should make sure that the `///<refrence`  is only emitted if --out or --flaten is used to external .d.ts and not to other source files that are part of the output.